From 3baf9b93dc1dfe38585722c71d7268304cb4a01a Mon Sep 17 00:00:00 2001
From: Alexander Bokovoy <abokovoy@redhat.com>
Date: Sun, 21 Sep 2025 11:14:51 +0300
Subject: libkrb5: in case PKINIT is configured, attempt Anonymous
 PKINIT for FAST

If auto_fast_armor is configured for the realm or globally, optimistically
assume that Anonymous PKINIT is supported as well and try to obtain it for
FAST use in case no pre-made FAST channel was established by the caller.

This behavior will automatically enable use of passwordless pre-authentication
methods which rely on FAST channel presence in deployments such as FreeIPA.

Notably, Microsoft Active Directory KDCs do not support Anonymous PKINIT. For
these deployments only a machine account (host keytab) can be used to build a
FAST channel. However, libkrb5 does not have access to /etc/krb5.keytab in a
general case.

Signed-off-by: Alexander Bokovoy <abokovoy@redhat.com>
---
 src/lib/krb5/krb/fast.c | 118 ++++++++++++++++++++++++++++++++++++++++
 src/lib/krb5/krb/fast.h |   2 +
 src/man/krb5.conf.man   |  13 +++++
 3 files changed, 133 insertions(+)

diff --git a/src/lib/krb5/krb/fast.c b/src/lib/krb5/krb/fast.c
index 62c9f0841..ee2e08189 100644
--- a/src/lib/krb5/krb/fast.c
+++ b/src/lib/krb5/krb/fast.c
@@ -168,6 +168,109 @@ krb5int_fast_prep_req_body(krb5_context context,
     return retval;
 }
 
+static krb5_boolean
+fast_is_pkinit_allowed(krb5_context context, krb5_data *realm)
+{
+    int value;
+    krb5_error_code retval = EINVAL;
+    char realmstr[1024];
+    const char *option = "auto_fast_armor";
+    const int def_value = FALSE;
+
+    if (realm != NULL && realm->length > sizeof(realmstr)-1)
+        return FALSE;
+
+    if (realm != NULL) {
+        strncpy(realmstr, realm->data, realm->length);
+        realmstr[realm->length] = '\0';
+
+        retval = profile_get_boolean(context->profile,
+                                     KRB5_CONF_REALMS, realmstr,
+                                     option, def_value, &value);
+    }
+
+    return retval ? FALSE : value;
+
+}
+
+static krb5_error_code
+fast_acquire_pkinit_armor(krb5_context context,
+                           struct krb5int_fast_request_state *state,
+                           krb5_get_init_creds_opt *opt, krb5_kdc_req *request)
+{
+    krb5_context ctx;
+    krb5_get_init_creds_opt *options = NULL;
+    krb5_error_code retval = 0;
+    krb5_data *target_realm = &request->server->realm;
+    krb5_creds creds;
+    krb5_principal anon_princ = NULL;
+    krb5_ccache out_cc;
+
+    /* short circuit, we are asked to perform Anonymous PKINIT already */
+    if (opt->flags & KRB5_GET_INIT_CREDS_OPT_ANONYMOUS) {
+        return EINVAL;
+    }
+
+    /* skip realms which do not allow use of automated FAST armor */
+    if (!fast_is_pkinit_allowed(context, target_realm)) {
+        return EINVAL;
+    }
+
+    retval = krb5_init_context(&ctx);
+    if (retval != 0) {
+        return retval;
+    }
+    retval = krb5_get_init_creds_opt_alloc(ctx, &options);
+    if (retval != 0) {
+        goto cleanup;
+    }
+    krb5_get_init_creds_opt_set_anonymous(options, 1);
+    retval = krb5_cc_new_unique(ctx, "MEMORY", NULL, &out_cc);
+    if (retval != 0) {
+        goto cleanup;
+    }
+
+    retval = krb5_get_init_creds_opt_set_out_ccache(ctx, options, out_cc);
+    if (retval != 0) {
+        goto cleanup;
+    }
+
+    retval = krb5_build_principal_ext(ctx, &anon_princ,
+                                      target_realm->length, target_realm->data,
+                                      strlen(KRB5_WELLKNOWN_NAMESTR),
+                                      KRB5_WELLKNOWN_NAMESTR,
+                                      strlen(KRB5_ANONYMOUS_PRINCSTR),
+                                      KRB5_ANONYMOUS_PRINCSTR, 0);
+    if (retval != 0) {
+        goto cleanup;
+    }
+
+    retval = krb5_get_init_creds_password(ctx, &creds, anon_princ, 0,
+                                          NULL /* no prompter */, NULL,
+                                          0, NULL /* service name */,
+                                          options);
+    if (retval == 0) {
+        state->fast_state_flags |= KRB5INT_FAST_OWN_ARMOR;
+        state->armor_ccache = out_cc;
+    }
+cleanup:
+    if (retval != 0 && out_cc != NULL) {
+        (void) krb5_cc_destroy(ctx, out_cc);
+    }
+    if (retval == 0) {
+        krb5_free_cred_contents(ctx, &creds);
+    }
+    if (options != NULL) {
+        krb5_get_init_creds_opt_free(ctx, options);
+    }
+    if (anon_princ != NULL) {
+        krb5_free_principal(ctx, anon_princ);
+    }
+    krb5_free_context(ctx);
+
+    return retval;
+}
+
 krb5_error_code
 krb5int_fast_as_armor(krb5_context context,
                       struct krb5int_fast_request_state *state,
@@ -178,10 +281,20 @@ krb5int_fast_as_armor(krb5_context context,
     krb5_principal target_principal = NULL;
     krb5_data *target_realm;
     const char *ccname = k5_gic_opt_get_fast_ccache_name(opt);
+    char *fast_ccname = NULL;
     krb5_flags fast_flags;
 
     krb5_clear_error_message(context);
     target_realm = &request->server->realm;
+    if (ccname == NULL) {
+        retval = fast_acquire_pkinit_armor(context, state, opt, request);
+        if (retval == 0) {
+            retval = krb5_cc_get_full_name(context, state->armor_ccache, &fast_ccname);
+            if (retval == 0 && fast_ccname != NULL)
+                ccname = fast_ccname;
+        }
+        retval = 0;
+    }
     if (ccname != NULL) {
         TRACE_FAST_ARMOR_CCACHE(context, ccname);
         state->fast_state_flags |= KRB5INT_FAST_ARMOR_AVAIL;
@@ -220,6 +333,8 @@ krb5int_fast_as_armor(krb5_context context,
         krb5_cc_close(context, ccache);
     if (target_principal)
         krb5_free_principal(context, target_principal);
+    if (fast_ccname)
+        free(fast_ccname);
     return retval;
 }
 
@@ -615,6 +730,9 @@ krb5int_fast_free_state(krb5_context context,
     /*We are responsible for none of the store in the fast_outer_req*/
     krb5_free_keyblock(context, state->armor_key);
     krb5_free_fast_armor(context, state->armor);
+    if (state->fast_state_flags & KRB5INT_FAST_OWN_ARMOR) {
+        krb5_cc_destroy(context, state->armor_ccache);
+    }
     free(state);
 }
 
diff --git a/src/lib/krb5/krb/fast.h b/src/lib/krb5/krb/fast.h
index 7156ea203..e5fe8bd54 100644
--- a/src/lib/krb5/krb/fast.h
+++ b/src/lib/krb5/krb/fast.h
@@ -34,6 +34,7 @@ struct krb5int_fast_request_state {
     krb5_kdc_req fast_outer_request;
     krb5_keyblock *armor_key; /*non-null means fast is in use*/
     krb5_fast_armor *armor;
+    krb5_ccache armor_ccache;
     krb5_ui_4 fast_state_flags;
     krb5_ui_4 fast_options;
     krb5_int32 nonce;
@@ -41,6 +42,7 @@ struct krb5int_fast_request_state {
 
 #define KRB5INT_FAST_DO_FAST     (1l<<0)  /* Perform FAST */
 #define KRB5INT_FAST_ARMOR_AVAIL (1l<<1)
+#define KRB5INT_FAST_OWN_ARMOR   (1l<<2)
 
 krb5_error_code
 krb5int_fast_prep_req_body(krb5_context context,
diff --git a/src/man/krb5.conf.man b/src/man/krb5.conf.man
index d4caa2bd3..ac7649647 100644
--- a/src/man/krb5.conf.man
+++ b/src/man/krb5.conf.man
@@ -650,6 +650,19 @@ primary KDC, in case the user\(aqs password has just been changed, and
 the updated database has not been propagated to the replica
 servers yet.  New in release 1.19.
 .TP
+\fBauto_fast_armor\fP
+If this flag is true, then initial ticket request will use Anonymous
+PKINIT to protect the communication as a FAST channel in case an application
+did not provide its own FAST channel. This is useful for deployments where
+pre-authentication methods require use of the FAST channel, such as
+passwordless methods provided by FreeIPA. Microsoft Active Directory
+implementation of PKINIT does not support Anonymous PKINIT feature.
+As a result, \fIauto_fast_armor\fP defaults to false.
+.sp
+Use of \fIauto_fast_armor = true\fP requires properly configured PKINIT and
+WELLKNOWN/ANONYMOUS principal defined on the KDC side. Consult KDC documentation
+for details.
+.TP
 \fBv4_instance_convert\fP
 This subsection allows the administrator to configure exceptions
 to the \fBdefault_domain\fP mapping rule.  It contains V4 instances
-- 
2.51.0

